src/project.rs

```diff
@@ -0,0 +1,95 @@
+use std::error::Error;
+use std::ffi::OsString;
+use std::fs::{read_dir, DirEntry};
+use std::io::prelude::*;
+
+pub type ProjectError = String;
+pub type Proc<T> = Result<T, ProjectError>;
+
+pub fn discover(dir: &str) -> Proc<Vec<GitRepo>> {
+    match read_dir(dir) {
+        Err(error) => Err(error.description().to_string()),
+        Ok(files) => {
+            for file_result in files {
+                match file_result {
+                    Ok(file) => {
+                        let x = try!(git_repo(file));
+                        println!("Found path: {:?}", x);
+                    }
+                    Err(err) => {
+                        return Err(err.description().to_string());
+                    }
+                }
+            }
+            Err("Finished work".to_string())
+        },
+    }
+}
+
+// TODO: Better parameter type for representing files?
+pub fn git_repo(root: DirEntry) -> Proc<GitRepo> {
```

Comment:
Well, flatMap over a `Result` is easy because it is "biased" which means it favors one side in all of its `map` like methods. I'm only referring to the fact that `filter_map` requires an `A => Option<B>`, so we can just ignore errors while reading files to get a smaller subset of directories that could be found. We might want some sort of recovery mechanism for stats that live on a `GitRepo` that no longer exists or can't be read for some reason, but that code doesn't belong here.